### PR TITLE
Adding XZ decompression functionality for skiboot

### PR DESCRIPTION
**Pull request by e-liner**
We have already merged hostboot and open-power/pnor changes, this is the final change for RTC:125550.

This change will allow a user the option to XZ compress the skiboot image. The BR2_TARGET_SKIBOOT_XZ config will need to be enabled for the specific configuration, and then the skiboot image will be compressed. This change is only for compressing and decompressing the skiboot image section. If another section needs compression, we would have to enable decompression inside hostboot for that specific section, and make additional pnor changes.

---
This change is [Reviewable](https://reviewable.io/reviews/open-power/op-build/418)

### CODE REVIEW

**Comment by williamspatrick** [comment]
@stewart-ibm - Please take a look at this and let me know if you have any concerns.  LGTM.

File: openpower/package/openpower-pnor/Config.in [hunk 1]
```diff
@@ -35,6 +35,17 @@ config BR2_SKIBOOT_LID_NAME
...
+config BR2_TARGET_SKIBOOT_XZ
+        boolean "Compress the skiboot image with XZ"
+        default n
+
+config BR2_SKIBOOT_LID_XZ_NAME
+        string "Name of compressed skiboot lid name to be used"
+        default "skiboot.lid" if !BR2_TARGET_SKIBOOT_XZ
+        default "skiboot.lid.xz" if BR2_TARGET_SKIBOOT_XZ
+        help
+            String used to define compressed sapphire lid filename
```
Comment (stewart-ibm) [inline openpower/package/openpower-pnor/Config.in]: should be skiboot rather than Sapphire.

Sapphire was only ever an internal name.

**Comment by stewart-ibm** [comment]
It looks like we don't enable this by default though... and I'm not sure why we wouldn't.

My concerns are mainly around the PNOR ffs format and upgrades. It seems that the flag in the partition header needs to be flipped, which means we may not be able to do upgrades on existing systems (and I don't see why we wouldn't want to... xz compressed skiboot just decreases boot time, and that's always a plus). We also need to update libflash and pflash to parse the new flag and flip it / abort accordingly.

I'd certainly prefer the HB code to test for a xz magic number if the partition is flagged as XZ compressable, as that makes it much easier for development as people will otherwise have to update their workflows.

In fact... we don't really need the flag in the partition at all, which means we avoid all of the upgrade pain. We just need to check the magic number at the start of the PAYLOAD partition, if it's the XZ header, then we go unxz it.

**Comment by dcrowell77** [comment]
I think you answered your own question on why this isn't enabled by default - the upgrade scenario.  There are a variety of issues to thoughtfully consider before changing any behavior that have previously been relied upon.  But putting the support in place allows any new systems to take advantage of the functionality.

Why would pflash be affected by this flag at all?  The data within the partition is just an opaque blob, whether or not it is compressed.  Or are you considering a switch from uncompressed to compressed without applying a new TOC?

Re: magic number vs ffs flag -- I guess I can see where you are coming from here.  I'm not familiar enough with the xz format to know how reliable looking at a magic number might be.  The support isn't in place yet, but we've considered making this a more generic platform for any kind of compression going forward, maybe that possibility warrants an explicit flag since not all formats would have a magic number?

**Comment by e-liner** [comment]
In the header of xz there's always "const uint8_t HEADER_MAGIC[6] = { 0xFD, '7', 'z', 'X', 'Z', 0x00 };", and that's the same for all xz compressed blobs.  My only concern with looking for the magic number is that this works for all xz compressed blobs and not our specific XZ options.  But, if we always create the pnor, then that's not a problem.  

I don't think it's likely that another compression algorithm will have the same magic number, but I suppose it is possible and that would definitely cause the decompression to fail.  I think that most compression algorithms have a magic number in the header, but the format isn't always the same; some algorithms start with the magic number in the header, and some have some buffer room before the magic number.

**Comment by e-liner** [comment]
Sorry about all of the updates, my environment keeps getting messed up and the rebasing is causing old commits to update and be seen as "new".  I've opened up a new clean pull request here: https://github.com/open-power/op-build/pull/425

**Comment by williamspatrick** [comment]
@stewart-ibm : Thanks for the feedback.  As Dan mentioned we are not
going to enable by default because of the concerns with backwards
compatibility.  You gave an idea (the magic number) on how to solve that
without needing any PNOR flags.

@e-liner has a follow up commit she needs to do anyhow.  What we are
going to do is:

1) Merge this change in, but do not enable via config flags.  This
allows people to try out the XZ code.

2) Have Elizabeth make the proposed changes to use the magic and thus
revert the PNOR flags.

3) Enable compression on all targets with the PNOR flag removal code up
in github.

On Wed, Mar 30, 2016 at 07:36:28PM -0700, Stewart Smith wrote:
> It looks like we don't enable this by default though... and I'm not sure why we wouldn't.
> 
> My concerns are mainly around the PNOR ffs format and upgrades. It seems that the flag in the partition header needs to be flipped, which means we may not be able to do upgrades on existing systems (and I don't see why we wouldn't want to... xz compressed skiboot just decreases boot time, and that's always a plus). We also need to update libflash and pflash to parse the new flag and flip it / abort accordingly.
> 
> I'd certainly prefer the HB code to test for a xz magic number if the partition is flagged as XZ compressable, as that makes it much easier for development as people will otherwise have to update their workflows.
> 
> In fact... we don't really need the flag in the partition at all, which means we avoid all of the upgrade pain. We just need to check the magic number at the start of the PAYLOAD partition, if it's the XZ header, then we go unxz it.
> 
> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/open-power/op-build/pull/418#issuecomment-203723328

-- 
Patrick Williams

**Comment by stewart-ibm** [comment]
Dan <notifications@github.com> writes:
> Why would pflash be affected by this flag at all?  The data within the
> partition is just an opaque blob, whether or not it is compressed.  Or
> are you considering a switch from uncompressed to compressed without
> applying a new TOC?

Without applying a new TOC, yes. Very very very common in development.

Requiring a new TOC would break development workflow for literally
everyone working on skiboot.

-- 
Stewart Smith
OPAL Architect, IBM.

**Comment by stewart-ibm** [comment]
Patrick Williams <notifications@github.com> writes:
> @stewart-ibm : Thanks for the feedback.  As Dan mentioned we are not
> going to enable by default because of the concerns with backwards
> compatibility.  You gave an idea (the magic number) on how to solve that
> without needing any PNOR flags.
>
> @e-liner has a follow up commit she needs to do anyhow.  What we are
> going to do is:
>
> 1) Merge this change in, but do not enable via config flags.  This
> allows people to try out the XZ code.
>
> 2) Have Elizabeth make the proposed changes to use the magic and thus
> revert the PNOR flags.
>
> 3) Enable compression on all targets with the PNOR flag removal code up
> in github.

Perfect!

looking forward to it (and the likely 0.2 seconds it'll shave off boot time)

-- 
Stewart Smith
OPAL Architect, IBM.

**Pull request closed without merging**